File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n=len(prices)
        next_greater=[0]*n
        next_greater[n-1]=-1
        for i in range(n-2,-1,-1):
            next_greater[i]=max(prices[i+1],next_greater[i+1])
        ans=0
        for i in range(0,n,1):
            ans=max(ans,next_greater[i]-prices[i])
        return ans

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int solve(int ind,int n, int tag,vector<int> &prices,vector<vector<int>> &dp){
        if(ind==n-1){
            if(tag==0) return 0;
            return prices[ind];
        }
        if(dp[ind][tag]!=-1) return dp[ind][tag];
        if(tag==0){
            int take=solve(ind+1,n,!tag,prices,dp)-prices[ind];
            int not_take=solve(ind+1,n,tag,prices,dp);
            return dp[ind][tag]=max(take,not_take);
        }else  return dp[ind][tag]=max(prices[ind],solve(ind+1,n,tag,prices,dp));
    }
    int maxProfit(vector<int>& prices) {
        int n=prices.size();
        if(n==1) return 0;
        vector<vector<int>> dp(n,vector<int>(2,-1));
        return solve(0,n,0,prices,dp);
    }
};